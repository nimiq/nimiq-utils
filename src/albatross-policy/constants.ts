/**
 * The total supply of the cryptocurrency, in LUNA.
 */
export const TOTAL_SUPPLY = 21e14;

/**
 * The date of the proof-of-stake migration.
 */
// export const PROOF_OF_STAKE_MIGRATION_DATE = new Date('2024-11-19T16:45:20.000Z')
export const PROOF_OF_STAKE_MIGRATION_DATE = new Date('2024-11-19');

/**
 * The total supply of the cryptocurrency at the proof-of-stake migration date, in NIM.
 *
 * Same as:
 * powSupplyAfter(powBlockHeightAt(PROOF_OF_STAKE_MIGRATION_DATE))
 */
export const SUPPLY_AT_PROOF_OF_STAKE_MIGRATION_DATE = 12_893_109_654.06244;

/**
 * The total supply of the cryptocurrency in **Testnet** at the proof-of-stake migration date, in NIM.
 *
 * Same as:
 * powSupplyAfter(powBlockHeightAt(PROOF_OF_STAKE_MIGRATION_DATE))
 */
export const SUPPLY_AT_PROOF_OF_STAKE_MIGRATION_DATE_TESTNET = 12_030_755_339.52899;

/**
 * The block height of the proof-of-stake migration.
 */
export const PROOF_OF_STAKE_MIGRATION_BLOCK = 3456000;

/**
 * The initial supply of the cryptocurrency, in Lunas.
 */
export const PROOF_OF_WORK_INITIAL_SUPPLY = 2520000000e5;

/**
 * The speed of the emission curve, in Lunas.
 */
export const PROOF_OF_WORK_EMISSION_SPEED = 4_194_304; // same as powi(2, 22)

/**
 * The block number at which the emission tail starts.
 */
export const PROOF_OF_WORK_EMISSION_TAIL_START = 48692960;

/**
 * The reward per block in the emission tail.
 */
export const PROOF_OF_WORK_EMISSION_TAIL_REWARD = 4000;

export const MAX_SIZE_MICRO_BODY: number = 100_000;
export const MAX_TX_SENDER_DATA_SIZE: number = 1;
export const MAX_TX_RECIPIENT_DATA_SIZE: number = 2112;
export const MAX_BASIC_TX_RECIPIENT_DATA_SIZE: number = 64;
export const MAX_MERKLE_PATH_SIZE: number = 1029;
export const MAX_SUPPORTED_WEB_AUTH_SIZE: number = 512;
export const VERSION: number = 1;
export const SLOTS: number = 512;
export const TWO_F_PLUS_ONE: number = Math.ceil((2 * SLOTS) / 3);
export const F_PLUS_ONE: number = Math.ceil(SLOTS / 3);
export const MIN_PRODUCER_TIMEOUT: number = 4 * 1000;
export const BLOCK_SEPARATION_TIME: number = 1000;
export const TENDERMINT_TIMEOUT_INIT: number = 4 * 1000;
export const TENDERMINT_TIMEOUT_DELTA: number = 1000;
export const MIN_EPOCHS_STORED: number = 1;
export const TIMESTAMP_MAX_DRIFT: number = 600000;
export const BLOCKS_DELAY_DECAY: number = 0.9999999989;
export const MINIMUM_REWARDS_PERCENTAGE: number = 0.5;
export const VALIDATOR_DEPOSIT: number = 10_000_000_000;
export const MINIMUM_STAKE: number = 10_000_000;
export const JAIL_EPOCHS: number = 8;
export const SUPPLY_DECAY: number = 0.9999999999960264;
export const BLS_CACHE_MAX_CAPACITY: number = 1000;
export const HISTORY_CHUNKS_MAX_SIZE: number = 25 * 1024 * 1024;
export const POS_DECAY_PER_DAY = 0.0003432600460362;
