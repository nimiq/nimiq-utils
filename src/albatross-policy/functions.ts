import {
    BLOCKS_DELAY_DECAY,
    BLOCKS_PER_BATCH,
    BLOCKS_PER_EPOCH,
    JAIL_EPOCHS,
    MINIMUM_REWARDS_PERCENTAGE,
    PROOF_OF_STAKE_MIGRATION_BLOCK,
    PROOF_OF_STAKE_MIGRATION_BLOCK_TESTNET,
    PROOF_OF_STAKE_MIGRATION_DATE,
    PROOF_OF_STAKE_MIGRATION_DATE_TESTNET,
} from './constants';

function getNetwork(network?: string) {
    if (!network) return { isMainnet: true, isTestnet: false }; // default to mainnet
    if (network?.toLowerCase().includes('test')) return { isMainnet: false, isTestnet: true };
    if (network?.toLowerCase().includes('main')) return { isMainnet: true, isTestnet: false };
    console.warn(`Network "${network}" not implemented.`);
    return { isMainnet: false, isTestnet: false };
}

export interface BaseAlbatrossPolicyOptions {
    /**
     * Whether to use the testnet block height for proof-of-stake migration.
     * Can be a boolean or a string containing 'test' to use testnet values.
     * @default 'main-albatross'
     */
    network?: string;
    /**
     * Custom migration block to use instead of the default.
     */
    migrationBlock?: number;
}

/* eslint-disable max-len */
/**
 * Gets the appropriate migration block based on options.
 * If a custom migrationBlock is provided in options, that takes precedence.
 * Otherwise, it checks if the network is testnet and returns the corresponding migration block.
 *
 * Check https://github.com/nimiq/core-rs-albatross/blob/7b6cb0f5f90afff74ccaa14fe33dec3622ce228e/primitives/src/networks.rs#L74-L85 to see the network names.
 */
export function getMigrationBlock(options: BaseAlbatrossPolicyOptions = { network: 'main-albatross' }): number {
    if (options.migrationBlock !== undefined) return options.migrationBlock;
    const { isMainnet, isTestnet } = getNetwork(options.network);
    if (isTestnet) return PROOF_OF_STAKE_MIGRATION_BLOCK_TESTNET;
    if (isMainnet) return PROOF_OF_STAKE_MIGRATION_BLOCK;
    throw new Error(`Network "${options.network}" not implemented.`);
}
/* eslint-disable-next-line max-len */

/**
 * Returns information about the migration block based on the network.
 */
export function getMigrationBlockInfo({ network }: Pick<BaseAlbatrossPolicyOptions, 'network'> = {}) {
    const migrationBlock = getMigrationBlock({ network });
    const { isTestnet, isMainnet } = getNetwork(network);
    if (!isTestnet && !isMainnet) {
        throw new Error(`Network "${network}" not implemented.`);
    }
    const date = isTestnet ? PROOF_OF_STAKE_MIGRATION_DATE_TESTNET : PROOF_OF_STAKE_MIGRATION_DATE;
    const timestamp = date.getTime();
    return { timestamp, date, migrationBlock };
}

/**
 * Returns the epoch number for a given block number.
 */
export function epochAt(blockNumber: number, options: BaseAlbatrossPolicyOptions = {}): number {
    const migrationBlock = getMigrationBlock(options);
    if (blockNumber <= migrationBlock) return 0;
    const offset = blockNumber - migrationBlock;
    return Math.floor((offset + BLOCKS_PER_EPOCH - 1) / BLOCKS_PER_EPOCH);
}

/**
 * Returns the epoch index for a given block number.
 * Adjusted to add BLOCKS_PER_EPOCH - 1 before taking the modulo.
 */
export function epochIndexAt(blockNumber: number, options: BaseAlbatrossPolicyOptions = {}): number {
    const migrationBlock = getMigrationBlock(options);
    if (blockNumber < migrationBlock) return blockNumber;
    return (blockNumber - migrationBlock + BLOCKS_PER_EPOCH - 1) % BLOCKS_PER_EPOCH;
}

/**
 * Returns the batch number for a given block number.
 */
export function batchAt(blockNumber: number, options: BaseAlbatrossPolicyOptions = {}): number {
    const migrationBlock = getMigrationBlock(options);
    if (blockNumber <= migrationBlock) return 0;
    const offset = blockNumber - migrationBlock;
    return Math.floor((offset + BLOCKS_PER_BATCH - 1) / BLOCKS_PER_BATCH);
}

/**
 * Returns the batch index for a given block number.
 * Adjusted to add BLOCKS_PER_BATCH - 1 before taking the modulo.
 */
export function batchIndexAt(blockNumber: number, options: BaseAlbatrossPolicyOptions = {}): number {
    const migrationBlock = getMigrationBlock(options);
    if (blockNumber < migrationBlock) return blockNumber;
    return (blockNumber - migrationBlock + BLOCKS_PER_BATCH - 1) % BLOCKS_PER_BATCH;
}

/**
 * Checks if a given block number is a macro block.
 */
export function isMacroBlockAt(blockNumber: number, options: BaseAlbatrossPolicyOptions = {}): boolean {
    const migrationBlock = getMigrationBlock(options);
    if (blockNumber < migrationBlock) return false;
    return batchIndexAt(blockNumber, options) === BLOCKS_PER_BATCH - 1;
}

/**
 * Checks if a given block number is an election block.
 */
export function isElectionBlockAt(blockNumber: number, options: BaseAlbatrossPolicyOptions = {}): boolean {
    const migrationBlock = getMigrationBlock(options);
    if (blockNumber < migrationBlock) return false;
    return epochIndexAt(blockNumber, options) === BLOCKS_PER_EPOCH - 1;
}

/**
 * Checks if a given block number is a micro block.
 */
export function isMicroBlockAt(blockNumber: number, options: BaseAlbatrossPolicyOptions = {}): boolean {
    const migrationBlock = getMigrationBlock(options);
    if (blockNumber < migrationBlock) return false;
    return !isMacroBlockAt(blockNumber, options);
}

/**
 * Returns the next macro block number after the given block number.
 */
export function macroBlockAfter(blockNumber: number, options: BaseAlbatrossPolicyOptions = {}): number | undefined {
    const migrationBlock = getMigrationBlock(options);
    if (blockNumber < migrationBlock) return migrationBlock;

    const offset = blockNumber - migrationBlock;
    const batchCount = Math.floor(offset / BLOCKS_PER_BATCH) + 1;

    // Check for safe integer multiplication
    const mulResult = batchCount * BLOCKS_PER_BATCH;
    if (!Number.isSafeInteger(mulResult)) return undefined;

    // Check for safe integer addition
    const addResult = mulResult + migrationBlock;
    if (!Number.isSafeInteger(addResult)) return undefined;

    return addResult;
}

/**
 * Returns the last macro block number before the given block number.
 */
export function lastMacroBlock(blockNumber: number, options: BaseAlbatrossPolicyOptions = {}): number | undefined {
    const migrationBlock = getMigrationBlock(options);
    if (blockNumber < migrationBlock) throw new Error('No macro blocks before proof-of-stake migration');

    const offset = blockNumber - migrationBlock;
    const batchCount = Math.floor(offset / BLOCKS_PER_BATCH);

    // Check for safe integer multiplication
    const mulResult = batchCount * BLOCKS_PER_BATCH;
    if (!Number.isSafeInteger(mulResult)) return undefined;

    // Check for safe integer addition
    const addResult = mulResult + migrationBlock;
    if (!Number.isSafeInteger(addResult)) return undefined;

    return addResult;
}

/**
 * Returns the next election block number after the given block number.
 */
export function electionBlockAfter(blockNumber: number, options: BaseAlbatrossPolicyOptions = {}): number | undefined {
    const migrationBlock = getMigrationBlock(options);
    if (blockNumber < migrationBlock) return migrationBlock;

    const offset = blockNumber - migrationBlock;
    const epochCount = Math.floor(offset / BLOCKS_PER_EPOCH) + 1;

    // Check for safe integer multiplication
    const mulResult = epochCount * BLOCKS_PER_EPOCH;
    if (!Number.isSafeInteger(mulResult)) return undefined;

    // Check for safe integer addition
    const addResult = mulResult + migrationBlock;
    if (!Number.isSafeInteger(addResult)) return undefined;

    return addResult;
}

/**
 * Returns the first block number of the given epoch.
 */
export function firstBlockOf(epoch: number, options: BaseAlbatrossPolicyOptions = {}): number | undefined {
    if (epoch <= 0) return undefined;

    const migrationBlock = getMigrationBlock(options);

    // Check for safe integer subtraction
    const subResult = epoch - 1;
    if (!Number.isSafeInteger(subResult)) return undefined;

    // Check for safe integer multiplication
    const mulResult = subResult * BLOCKS_PER_EPOCH;
    if (!Number.isSafeInteger(mulResult)) return undefined;

    // Check for safe integer addition
    const addResult = mulResult + migrationBlock + 1;
    if (!Number.isSafeInteger(addResult)) return undefined;

    return addResult;
}

/**
 * Returns the block number after the jail period.
 */
export function blockAfterJail(blockNumber: number): number | undefined {
    const mulResult = BLOCKS_PER_EPOCH * JAIL_EPOCHS;
    if (!Number.isSafeInteger(mulResult)) return undefined;
    const addResult = blockNumber + mulResult + 1;
    if (!Number.isSafeInteger(addResult)) return undefined;
    return addResult;
}

/**
 * Returns the election block number for a given epoch.
 */
export function electionBlockOf(epoch: number, options: BaseAlbatrossPolicyOptions = {}): number | undefined {
    if (epoch < 0) return undefined;
    const genesisBlock = getMigrationBlock(options);
    const mulResult = epoch * BLOCKS_PER_EPOCH;
    if (!Number.isSafeInteger(mulResult)) return undefined;
    const addResult = mulResult + genesisBlock;
    if (!Number.isSafeInteger(addResult)) return undefined;
    return addResult;
}

/**
 * Calculates rewards penalty due to batch delay.
 */
export function batchDelayPenalty(delayMs: number): number {
    if (!Number.isSafeInteger(delayMs) || delayMs < 0) {
        return MINIMUM_REWARDS_PERCENTAGE;
    }

    // Calculate (BLOCKS_DELAY_DECAY ** delayMs)
    const decayPower = BLOCKS_DELAY_DECAY ** delayMs;
    // Calculate final decay = decayPower ** delayMs
    const finalDecay = decayPower ** delayMs;

    return (1 - MINIMUM_REWARDS_PERCENTAGE) * finalDecay + MINIMUM_REWARDS_PERCENTAGE;
}

/**
 * Returns true if the block number is the first batch of an epoch.
 */
export function firstBatchOfEpoch(blockNumber: number, options: BaseAlbatrossPolicyOptions = {}): boolean {
    return epochIndexAt(blockNumber, options) < BLOCKS_PER_BATCH;
}

/**
 * Returns the last block number of the reporting window for a given block number.
 */
export function lastBlockOfReportingWindow(blockNumber: number): number | undefined {
    const addResult = blockNumber + BLOCKS_PER_EPOCH;
    if (!Number.isSafeInteger(addResult)) return undefined;
    return addResult;
}

/**
 * Returns the block number immediately after the reporting window for a given block number.
 */
export function blockAfterReportingWindow(blockNumber: number): number | undefined {
    const lastBlock = lastBlockOfReportingWindow(blockNumber);
    if (lastBlock === undefined) return undefined;

    const addResult = lastBlock + 1;
    if (!Number.isSafeInteger(addResult)) return undefined;

    return addResult;
}
