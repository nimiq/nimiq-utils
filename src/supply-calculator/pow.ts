import {
    PROOF_OF_STAKE_MIGRATION_BLOCK,
    PROOF_OF_STAKE_MIGRATION_DATE,
    PROOF_OF_WORK_EMISSION_SPEED,
    PROOF_OF_WORK_EMISSION_TAIL_REWARD,
    PROOF_OF_WORK_EMISSION_TAIL_START,
    PROOF_OF_WORK_INITIAL_SUPPLY,
} from '../albatross-policy/albatross-policy';

function _powBlockRewardAt(currentSupply: number, blockHeight: number): number {
    if (blockHeight <= 0) return 0;
    const remaining = PROOF_OF_WORK_INITIAL_SUPPLY - currentSupply;
    if (blockHeight >= PROOF_OF_WORK_EMISSION_TAIL_START && remaining >= PROOF_OF_WORK_EMISSION_TAIL_REWARD) {
        return PROOF_OF_WORK_EMISSION_TAIL_REWARD;
    }
    const remainder = remaining % PROOF_OF_WORK_EMISSION_SPEED;
    return (remaining - remainder) / PROOF_OF_WORK_EMISSION_SPEED;
}

function _powSupplyAfter(initialSupply: number | undefined, blockHeight: number, startHeight = 0): number {
    let supply = initialSupply || 0;
    for (let i = startHeight; i <= blockHeight; ++i) {
        supply += _powBlockRewardAt(supply, i);
    }
    return supply;
}

/**
 * Extrapolate PoW block height for a certain date.
 */
export function powBlockHeightAt(date: Date): number {
    return PROOF_OF_STAKE_MIGRATION_BLOCK + ((date.getTime() - PROOF_OF_STAKE_MIGRATION_DATE.getTime()) / (60e3));
}

const supplyCache = new Map<number, number>();
const supplyCacheInterval = 5000; // blocks
let supplyCacheMax = 0; // blocks

/**
 * Calculate the PoW supply at a given time.
 * @returns {number} The total supply of the cryptocurrency at the given time, in NIM.
 */
export function powSupplyAfter(blockHeight: number): number {
    // Calculate last entry in supply cache that is below blockHeight.
    let startHeight = Math.floor(blockHeight / supplyCacheInterval) * supplyCacheInterval;
    startHeight = Math.max(0, Math.min(startHeight, supplyCacheMax));

    // Calculate respective block for the last entry of the cache and the targeted height.
    const startI = startHeight / supplyCacheInterval;
    const endI = Math.floor(blockHeight / supplyCacheInterval);

    // The starting supply is the initial supply at the beginning and a cached value afterwards.
    let supply = startHeight === 0 ? PROOF_OF_WORK_INITIAL_SUPPLY : supplyCache.get(startHeight);
    // Use and update cache.
    for (let i = startI; i < endI; ++i) {
        startHeight = i * supplyCacheInterval;
        // Since the cache stores the supply *before* a certain block, subtract one.
        const endHeight = (i + 1) * supplyCacheInterval - 1;
        supply = _powSupplyAfter(supply, endHeight, startHeight);
        // Don't forget to add one again.
        supplyCache.set(endHeight + 1, supply);
        supplyCacheMax = endHeight + 1;
    }

    // Calculate remaining supply (this also adds the block reward for endI*interval).
    return _powSupplyAfter(supply, blockHeight, endI * supplyCacheInterval) / 1e5; // Luna >> NIM
}
