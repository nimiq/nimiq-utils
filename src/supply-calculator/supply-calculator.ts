export * from './pos';
export * from './pow';
export * from './utils';
