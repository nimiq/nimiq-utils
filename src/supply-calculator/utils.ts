/* eslint-disable max-len */

/**
 * Adapted `exp_by_squaring_iterative` from
 * https://en.wikipedia.org/w/index.php?title=Exponentiation_by_squaring&amp%3Boldid=1229001691&useskin=vector#With_constant_auxiliary_memory
 *
 * The Rust implementation is also using this algorithm to guarantee the same result on different platforms.
 */
export function powi(x: number, n: number): number {
    if (n < 0) {
        x = 1 / x;
        n *= -1;
    }
    if (!n) return 1;
    let y = 1;
    while (n > 1) {
        if (n % 2) {
            y *= x;
            n -= 1;
        }
        x *= x;
        n /= 2;
    }
    return x * y;
}
