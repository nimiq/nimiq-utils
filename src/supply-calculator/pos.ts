import {
    PROOF_OF_STAKE_MIGRATION_DATE,
    SUPPLY_AT_PROOF_OF_STAKE_MIGRATION_DATE,
    SUPPLY_AT_PROOF_OF_STAKE_MIGRATION_DATE_TESTNET,
    SUPPLY_DECAY,
    TOTAL_SUPPLY,
} from '../albatross-policy/albatross-policy';
import { powi } from './utils';

export interface PosSupplyAtParams {
    /**
     * The network name
     *
     * @default 'main-albatross'
     */
    network?: 'main-albatross' | 'test-albatross'
}

/**
 * Calculate the PoS supply at a given time.
 * @param {number} timestampMs The timestamp at which to calculate the PoS supply.
 * @returns {number} The total supply of the cryptocurrency at the given time, in NIM.
 */
export function posSupplyAt(timestampMs: number, { network = 'main-albatross' }: PosSupplyAtParams = {}): number {
    const ts = timestampMs - PROOF_OF_STAKE_MIGRATION_DATE.getTime();
    if (ts < 0) {
        throw new Error(
            `Invalid timestamp: ${timestampMs}. It must be greater than ${PROOF_OF_STAKE_MIGRATION_DATE.getTime()}.`,
        );
    }

    const genesisSupply = network === 'test-albatross'
        ? SUPPLY_AT_PROOF_OF_STAKE_MIGRATION_DATE_TESTNET
        : SUPPLY_AT_PROOF_OF_STAKE_MIGRATION_DATE;
    return (TOTAL_SUPPLY - ((TOTAL_SUPPLY - genesisSupply * 1e5) * powi(SUPPLY_DECAY, ts))) / 1e5;
}
