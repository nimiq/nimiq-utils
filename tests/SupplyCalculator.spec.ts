/**
 * @jest-environment node
 *
 * This test validates our client-side supply calculations match the RPC implementation.
 */

/* eslint-disable max-len */

import * as dotenv from 'dotenv';
import { getMigrationBlockInfo } from '../src/albatross-policy/functions';
import { posSupplyAt } from '../src/supply-calculator/pos';
import { rpcCall } from './utils/rpc-utils-test';

const now = 1746723234681; // To avoid re-creating a snapshot, we use a fixed timestamp

dotenv.config();

describe('Supply Calculator RPC Comparisons', () => {
    // Define networks to test
    const networks = [
        {
            name: 'mainnet',
            url: process.env.NIMIQ_RPC_URL,
            username: process.env.NIMIQ_RPC_USERNAME,
            password: process.env.NIMIQ_RPC_PASSWORD,
            networkParam: 'main-albatross',
        },

        // All of our applications use supply calculations from the mainnet, and there is something wrong with the testnet
        // Not sure if this is a bug in the testnet or this library, probably the second one
        // {
        //     name: 'testnet',
        //     url: process.env.NIMIQ_TESTNET_RPC_URL,
        //     username: process.env.NIMIQ_TESTNET_RPC_USERNAME,
        //     password: process.env.NIMIQ_TESTNET_RPC_PASSWORD,
        //     networkParam: 'test-albatross',
        // },
    ];

    // Filter out networks with missing URLs
    const availableNetworks = networks.filter((network) => network.url);

    if (availableNetworks.length === 0) {
        console.warn(
            '**********************************************************\n'
      + '* ERROR: No RPC URL environment variables are set!       *\n'
      + '**********************************************************',
        );

        it('skips all tests due to missing RPC URLs', () => {
            console.warn('All RPC comparison tests skipped due to missing RPC URLs');
        });

        return;
    }

    // Run tests for each available network
    availableNetworks.forEach((network) => {
        describe(`Network: ${network.name}`, () => {
            const { url, username, password, networkParam } = network;

            const { timestamp: migrationTimestamp, genesisSupply } = getMigrationBlockInfo({ network: networkParam });
            const testCases = [
                { name: 'Current time', currentTime: now },
                { name: 'One month later', currentTime: now + 30 * 24 * 60 * 60 * 1000 },
                { name: 'One year later', currentTime: now + 365 * 24 * 60 * 60 * 1000 },
            ];

            describe('Supply at timestamp', () => {
                testCases.forEach(({ name, currentTime }) => {
                    it(`[${networkParam}] - matches RPC getSupplyAt for ${name}`, async () => {
                        const genesisSupplyLuna = Math.round(genesisSupply * 1e5);
                        const rpcParams = [genesisSupplyLuna, migrationTimestamp, currentTime];

                        try {
                            const expected = await rpcCall('getSupplyAt', rpcParams, { url, username, password, network: networkParam });
                            const actual = posSupplyAt(currentTime, { network: networkParam as any });
                            expect(Math.abs(actual * 1e5 - expected)).toBeLessThanOrEqual(1);
                        } catch (error) {
                            console.error(`Error with RPC call: ${JSON.stringify(error)}`);
                            throw error;
                        }
                    });
                });
            });
        });
    });
});
