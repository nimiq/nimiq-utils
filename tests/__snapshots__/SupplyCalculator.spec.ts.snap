// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Supply Calculator RPC Comparisons Network: mainnet Supply at timestamp [main-albatross] - matches RPC getSupplyAt for Current time: getSupplyAt([1289310965406244,1732034720000,1746723234681]) 1`] = `
{
  "method": "getSupplyAt",
  "params": [
    1289310965406244,
    1732034720000,
    1746723234681,
  ],
  "result": null,
}
`;

exports[`Supply Calculator RPC Comparisons Network: mainnet Supply at timestamp [main-albatross] - matches RPC getSupplyAt for Current time: getSupplyAt([1289310965406244,1732034720000,1746723234681]) 2`] = `
{
  "method": "getSupplyAt",
  "params": [
    1289310965406244,
    1732034720000,
    1746723234681,
  ],
  "result": "placeholder",
}
`;

exports[`Supply Calculator RPC Comparisons Network: mainnet Supply at timestamp [main-albatross] - matches RPC getSupplyAt for Current time: getSupplyAt([1289310965406244,1732034720000,1746723234681]) 3`] = `
{
  "method": "getSupplyAt",
  "params": [
    1289310965406244,
    1732034720000,
    1746723234681,
  ],
  "result": 1335273483335262,
}
`;

exports[`Supply Calculator RPC Comparisons Network: mainnet Supply at timestamp [main-albatross] - matches RPC getSupplyAt for One month later: getSupplyAt([1289310965406244,1732034720000,1749315234681]) 1`] = `
{
  "method": "getSupplyAt",
  "params": [
    1289310965406244,
    1732034720000,
    1749315234681,
  ],
  "result": null,
}
`;

exports[`Supply Calculator RPC Comparisons Network: mainnet Supply at timestamp [main-albatross] - matches RPC getSupplyAt for One month later: getSupplyAt([1289310965406244,1732034720000,1749315234681]) 2`] = `
{
  "method": "getSupplyAt",
  "params": [
    1289310965406244,
    1732034720000,
    1749315234681,
  ],
  "result": "placeholder",
}
`;

exports[`Supply Calculator RPC Comparisons Network: mainnet Supply at timestamp [main-albatross] - matches RPC getSupplyAt for One month later: getSupplyAt([1289310965406244,1732034720000,1749315234681]) 3`] = `
{
  "method": "getSupplyAt",
  "params": [
    1289310965406244,
    1732034720000,
    1749315234681,
  ],
  "result": 1343109414398216,
}
`;

exports[`Supply Calculator RPC Comparisons Network: mainnet Supply at timestamp [main-albatross] - matches RPC getSupplyAt for One year later: getSupplyAt([1289310965406244,1732034720000,1778259234681]) 1`] = `
{
  "method": "getSupplyAt",
  "params": [
    1289310965406244,
    1732034720000,
    1778259234681,
  ],
  "result": null,
}
`;

exports[`Supply Calculator RPC Comparisons Network: mainnet Supply at timestamp [main-albatross] - matches RPC getSupplyAt for One year later: getSupplyAt([1289310965406244,1732034720000,1778259234681]) 2`] = `
{
  "method": "getSupplyAt",
  "params": [
    1289310965406244,
    1732034720000,
    1778259234681,
  ],
  "result": "placeholder",
}
`;

exports[`Supply Calculator RPC Comparisons Network: mainnet Supply at timestamp [main-albatross] - matches RPC getSupplyAt for One year later: getSupplyAt([1289310965406244,1732034720000,1778259234681]) 3`] = `
{
  "method": "getSupplyAt",
  "params": [
    1289310965406244,
    1732034720000,
    1778259234681,
  ],
  "result": 1425341356325259,
}
`;
